RedHatInsights/rhsm-subscriptions
# SWATCH-1175: Implement customer authn/authz in swatch-contracts

Jira issue: [SWATCH-1175](https://issues.redhat.com/browse/SWATCH-1175)

Description
===========
Implement customer-facing API support in the swatch-contracts service.

I added support for a customer variant of the `x-rh-identity` header, and integrated RBAC service to
add the `customer` role when either `subscriptions:*:*` or `subscriptions:reports:read` permissions
are returned by the RBAC service.

In order to do so, there were a few incidental changes worth noting:
* Updated `swatch.quarkus-rest-client-conventions` to suppoprt `quarkus` subprojects.
* Created a quarkus-specific subproject for the RBAC client (`project(":clients:quarkus:rbac-client")`)
* Added `x-rh-identity` as a header parameter to our rbac spec copy (this simplified integration in
  quarkus)
* Copied the capacity metrics definition from the customer-facing API spec into swatch-contracts.
  * I tried referencing it via `$ref`, but this caused problems including swatch-contracts having
    an incorrect openapi spec at runtime.
* Added a test to ensure swatch-contracts API definitions that are for customer facing APIs are
  equivalent to the customer-facing API spec, except for security definitions and error response
  definitions. (These are changes that could be applied to the monolith spec, but that felt
  like too much scope creep).
  * The security pattern we're using in swatch-contracts doesn't match the customer-facing API.
  * I found a way to simplify the error response definition without using `$ref`.

I also found a couple of quirks in smallrye openapi handling and went ahead and addressed them:
* By setting `mp.openapi.scan.exclude.packages=com.redhat.swatch.contract.openapi`, the generated
  API classes are not merged with the static file that defined them (this was causing duplicate
  incorrect enums - uppercase variants that weren't actually valid).

RBAC service integration configuration:
* `RBAC_ENABLED` enables the integration; defaults to true, except in dev mode.
* `RBAC_ENDPOINT` loaded from clowder config, defaulting to http://localhost:8080 if clowder
  provided config is unavailable.

Testing
=======

Setup
-----

1. Run the service with authorization and RBAC turned on:
```shell
cd swatch-contracts
RBAC_ENABLED=true QUARKUS_SECURITY_AUTH_ENABLED_IN_DEV_MODE=true quarkus dev
```

Steps
-----

See that if no auth is provided the endpoint returns 401:
1. Hit the endpoint without auth, see `401 Unauthorized` response:
```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/RHEL/Sockets \
 beginning==2022-01-01T00:00Z \
 ending==2022-01-01T00:00Z \
 granularity==DAILY
```

2. Try the endpoint with proper auth, see `403 Forbidden` because the RBAC service call failed
   with a warning in the logs:

```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/RHEL/Sockets \
 beginning==2022-01-01T00:00Z \
 ending==2022-01-01T00:00Z \
 granularity==DAILY \
 x-rh-identity:$(echo '{
  "identity":{
    "type":"User",
    "org_id": "org123"
  }}' | base64 -w0)
```

3. Set up a mock rbac service that gives the swatch admin role:
```shell
echo "HTTP/1.1 200 OK" > /tmp/mockrbac.txt
echo "Content-Type: application/json" >> /tmp/mockrbac.txt
echo >> /tmp/mockrbac.txt
echo '{"data":[{"permission":"subscriptions:*:*"}]}' >> /tmp/mockrbac.txt
nc -l 8080 -k -c "cat /tmp/mockrbac.txt"
```

4. Hit the endpoint with a non-customer x-rh-identity header, see `403 Forbidden` without
   interaction with the RBAC service:
```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/RHEL/Sockets \
 beginning==2022-01-01T00:00Z \
 ending==2022-01-01T00:00Z \
 granularity==DAILY \
 x-rh-identity:$(echo '{
  "identity":{
    "type":"Associate",
    "associate" : {
      "email": "test@example.com"
    }
  }}' | base64 -w0)
```

5. Try the customer identity again, this time see `500 Internal Server Error` w/
   `Not yet implemented`:

```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/RHEL/Sockets \
 beginning==2022-01-01T00:00Z \
 ending==2022-01-01T00:00Z \
 granularity==DAILY \
 x-rh-identity:$(echo '{
  "identity":{
    "type":"User",
    "org_id": "org123"
  }}' | base64 -w0)
```

6. Change the mock service to provide no permissions:
```shell
echo "HTTP/1.1 200 OK" > /tmp/mockrbac.txt
echo "Content-Type: application/json" >> /tmp/mockrbac.txt
echo >> /tmp/mockrbac.txt
echo '{"data":[]}' >> /tmp/mockrbac.txt
```

7. Try the customer API again, see `403 Forbidden` this time.
```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/RHEL/Sockets \
 beginning==2022-01-01T00:00Z \
 ending==2022-01-01T00:00Z \
 granularity==DAILY \
 x-rh-identity:$(echo '{
  "identity":{
    "type":"User",
    "org_id": "org123"
  }}' | base64 -w0)
```

8. Change the mock service to provide the swatch reader permission:
```shell
echo "HTTP/1.1 200 OK" > /tmp/mockrbac.txt
echo "Content-Type: application/json" >> /tmp/mockrbac.txt
echo >> /tmp/mockrbac.txt
echo '{"data":[{"permission":"subscriptions:reports:read"}]}' >> /tmp/mockrbac.txt
```

9. Try the customer API again, see `500 Internal Server Error` w/ `Not yet implemented` this time.
```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/RHEL/Sockets \
 beginning==2022-01-01T00:00Z \
 ending==2022-01-01T00:00Z \
 granularity==DAILY \
 x-rh-identity:$(echo '{
  "identity":{
    "type":"User",
    "org_id": "org123"
  }}' | base64 -w0)
```
